### PR TITLE
Cerberus Validator : Validate() Method

### PR DESCRIPTION
**Pull request by vipulgupta2048**
Signed-off-by: Vipul Gupta (@vipulgupta2048) vipulgupta2048@gmail.com

I originally wanted this PR to be here.

### CODE REVIEW

**Comment by vipulgupta2048** [comment]
@rennerocha read your comment on the previous PR, will implement that.

**Comment by vipulgupta2048** [comment]
As per @ejulio request for having one PR per card, I have made the changes to this PR. The translator doesn't have all the messages (I am adding them as I write the tests) but enough to see it working.

Let me know if there are changes that need to be done. I will be writing the tests in my local but would be really convenient if this could be merged so I can carry on writing and committing in sync. Thanks :hatching_chick:

File: spidermon/contrib/validation/cerberus/validator.py
Comment (ejulio) [inline spidermon/contrib/validation/cerberus/validator.py:23]: Each argument could go in one line.
Just readability tips :wink: 
Also, since these are not used by this class, only the super class, they could be `*args` and **kwargs`

**Review by ejulio — COMMENTED** [review]
Hey @vipulgupta2048 , nice : work.
Don't take the "one PR per card" too serious, it is a kind of balance.
This is a kind of special case that we separated tests and and the validator, but they should be together.
Every new feature/fix should get the proper unit tests in the same PR.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:23]: Thanks, Julio, really need to follow on the best practices. I just formatted my code with `Black` and submitted it. Will fix it.
I have read about `*args` and `**kwargs`, let me try to implement what you are saying and get back to you for a re-review if I did it right.

**Comment by vipulgupta2048** [comment]
I will have the tests ready to be put in a PR just after this merges.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:23]: I read about `*args` and `**kwargs` and I think, these won't work out. 
1. Since we would need them to initialize the default_translator variable quite explicitly. 
2. I tried it and it doesn't work out. 
Resolving.

**Review by vipulgupta2048 — COMMENTED** [review]
I think this is good to go now, :rocket: 
Reviews requested @ejulio @rennerocha :hatching_chick:

**Comment by vipulgupta2048** [comment]
Also, I am keeping the tests for the validator and translator in different files. The structure we are following with multiple tests running together through a single loop can only run for one format of tests and testing two different components (translator + validator) will make things a lot more messy :boom: 
Pushing a commit for that.

**Review by vipulgupta2048 — COMMENTED** [review]
:heavy_check_mark: G2G

File: spidermon/contrib/validation/cerberus/validator.py
Comment (ejulio) [inline spidermon/contrib/validation/cerberus/validator.py:14]: @rennerocha , what do you think of absolute imports here?
Also, should we keep both as absolute imports or mixing absolute;relative is fine here?

File: spidermon/contrib/validation/cerberus/validator.py
Comment (ejulio) [inline spidermon/contrib/validation/cerberus/validator.py]: No need for this variable.
Just `for field_name, error in validator.errors.items():`

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (ejulio) [inline tests/contrib/validation/cerberus/test_validator_cerberus.py]: It is fine for now, but it would be nice to refactor it to use `pytest`

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (ejulio) [inline tests/contrib/validation/cerberus/test_validator_cerberus.py]: PascalCase for classes

**Review by ejulio — COMMENTED** [review]
Just minor comments @vipulgupta2048 .
I'll let @rennerocha give the final word here for approval.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:14]: I researched about absolute imports and almost always its recommended to go and stay with absolute. Hence, I felt it could be justified here. 

[Reference #1](https://docs.python.org/2/whatsnew/2.5.html#pep-328-absolute-and-relative-imports), [Reference #2](https://blog.tankywoo.com/python/2013/10/07/python-relative-and-absolute-import.html)

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py]: Good addition, coming right up.

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (vipulgupta2048) [reply tests/contrib/validation/cerberus/test_validator_cerberus.py]: Looking forward to that, adding a card for it.

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (vipulgupta2048) [reply tests/contrib/validation/cerberus/test_validator_cerberus.py]: Pascal Case Coming up.. :running_man:

**Comment by vipulgupta2048** [comment]
@rennerocha Review needed here.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py]: It is better to raise an exception if the schema is not provided as soon as possible. We shouldn't wait until calling `_validate`. I'd prefer this validation in `__init__`.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py]: According to [Cerberus documentation](https://docs.python-cerberus.org/en/stable/schemas.html#validation-schemas), a valid schema is a [mapping](https://docs.python.org/3/glossary.html#term-mapping), so the error message should comply with this rule to avoid confusion.

Comment (rennerocha) [reply spidermon/contrib/validation/cerberus/validator.py]: Checking if object is an instance of `collections.abc.Mapping` should be enough to verify if the schema provided has the right format (not that it is a valid schema).

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py:37]: This exception is in the right place IMHO, as the schema is a valid mapping, but when we try to validate a document against it, it fails because the schema is not valid.

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (rennerocha) [inline tests/contrib/validation/cerberus/test_validator_cerberus.py]: I don't like this meta-class way to define the tests. I prefer the tests to have more explicit intentions.
Also, as we are creating an entire new feature, we could use pytest from the beginning.


For example:
```
def test_schema_in_invalid_format_provided():
    with pytest.raises(DocumentError):
        validator = CerberusValidator("I_AM_NOT_A_MAPPING")
```

Create the other tests like functions. It would be easier to understand what is being tested.

**Review by rennerocha — CHANGES_REQUESTED** [review]

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py]: Completely agree with that. Should I work towardsraise a new error there or shift this only.

Here's something I have in mind. Borrowed from Cerberus Source. 

```
 if schema is not None:
            self._schema = schema
        elif self._schema is None:
            if isinstance(self._schema, Mapping):
                self._schema = {}
            else:
                raise SchemaError("validation schema missing")
```

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py]: Now, that I check out `Mapping`, I realize this is actually a better way to go!
Again, should I just change the error message (Let Cerberus check it) or should we bypass with our own method which is kinda the same way

I am going with changing the message, to something more simpler and easy to understand.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:37]: Ahhhh, I get it. That's really actually great. We should be checking the schema first, if invalid then no reason going further. Gotcha :100:

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (vipulgupta2048) [reply tests/contrib/validation/cerberus/test_validator_cerberus.py]: Renne, I am good with that. Creating functions for each test. But, if we could be able to refactor the loop approach that is working right now with unit tests in  PyTest. Then, it it would be really really efficient what say? 

I want to work towards that only. Otherwise, for this, I will just create their functions. 
Also, I think it would raise a `ValueError`, No? (In your code snippet)

**Comment by vipulgupta2048** [comment]
@rennerocha I have implemented your suggestions as I thought could be best implemented. If the changes look good, please do resolve the conversation or let me know what more should be added in this PR. Looking forward to the review.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [inline spidermon/contrib/validation/cerberus/validator.py]: Do let me know if this is something that would be good to have.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py]: This import will not work on Python 2.7. You need to find the equivalent import and check for the Python version here.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py]: If `self._schema is None`, `isinstance(self._schema, Mapping)` will always return `False`. So this code will never be executed.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [reply spidermon/contrib/validation/cerberus/validator.py]: This is fine.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py:36]: Only `validator.validate(data)` would raise a `SchemaError`. Don't put too many code inside try..except.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (rennerocha) [inline spidermon/contrib/validation/cerberus/validator.py:37]: You are raising `SchemaError` in `__init__`, so the validator will not be instantiated if the schema is not valid, so I believe it is unnecessary to catch the exception here as well (only if it is possible to get a SchemaError for a reason different for the one you check in __init__.

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (rennerocha) [inline tests/contrib/validation/cerberus/test_validator_cerberus.py]: All parameters are invalid schemas? So your test name should be more informative like `test_raise_value_error_with_invalid_schemas` or something like that.  `test_function` is not a good test name.

Comment (rennerocha) [reply tests/contrib/validation/cerberus/test_validator_cerberus.py]: Also, where are the tests with valid schemas? You need to verify if the happy-path is working too.

**Review by rennerocha — CHANGES_REQUESTED** [review]
Left more comments. Please review them.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py]: Good detail didn't think of backward compatibility since Python 2.7 is reaching EOL. Still, I think its a good step, will resolve this.

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:36]: Yep, that's true. Will get other lines out of there!

File: spidermon/contrib/validation/cerberus/validator.py
Comment (ejulio) [reply spidermon/contrib/validation/cerberus/validator.py]: Probably this `if` would be easier to read as
```
if schema:
   ...
else:
   ....
```

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:37]: Yes, Renne both errors are from the same error class apparently. That's why 2 catch, raise.
`>>> cerberus.schema.SchemaError: validation schema missing` and 
`>>> cerberus.schema.SchemaError: {'a': ['must be of dict type']}`

**Reasons**
- We am raising `SchemaError` from `__init__` because we are importing SchemaError from Cerberus , hence it only make to raise it when no schema is given. 

- On the other hand, we are catching the `SchemaError`, to raise ValueError instead for uniformity.

File: tests/contrib/validation/cerberus/test_validator_cerberus.py
Comment (vipulgupta2048) [reply tests/contrib/validation/cerberus/test_validator_cerberus.py]: That's on me, I think between deleting and stuff I might have misplaced the name you actually suggested in the first place. More descriptive test names coming right up. 

About Valid schemas, we can't actually do that without raising an assertion error, in this case, can we. I can introduce a property of `valid=True` but then we have to also have expected errors, as they go hand in hand with the output of validator if that works? I personally also like to verify the happy path but would like some insight on the best way of implementing it.

Comment (ejulio) [reply tests/contrib/validation/cerberus/test_validator_cerberus.py]: For a happy path, it would be a new test function.
`test_invalid_schema` and all it's test cases
`test_valid_schema` and all it's  test cases (if it applies for more than one test)

File: spidermon/contrib/validation/cerberus/validator.py
Comment (vipulgupta2048) [reply spidermon/contrib/validation/cerberus/validator.py:36]: On testing again, this also produces SchemaError, I think  that's why it was in the block in the first place, `validator = cerberus_validator(self._schema)`

Removed the error loop though,

**Review by ejulio — APPROVED** [review]
hey @vipulgupta2048 , it looks good to me.
Lets just wait @rennerocha comments

**Review by rennerocha — APPROVED** [review]
Looks good to me.

**Pull request merged**
